New Wizard Spell/Item: "Staff of Thunder"

Flaborized opened the pull request:
<!-- The text between the arrows are comments - they will not be visible on your PR. -->
<!-- To automatically tag this PR, add the uppercase label(s) surrounded by brackets below, for example: [LABEL] -->

[FEAT]

## About the PR <!-- Describe the Pull Request here. What does it change? What other things could this impact? -->
Not WIP anymore, I added what I wanted to add

This PR adds a new 2 cost wizard spell and item, the thunder stave. It's in some ways similar to the Staff of Cthulhu (SoC), but in other ways not similar at all.

- It costs two points and comes with a spell and item, just like SoC.
- Clicking on a tile within 10 tiles (including in the fog of war but *not* including tiles further away through cameras or pods or whatever) will summon an electrical looking target. After 3 seconds (subject to change) a bolt of lightning strikes the target, leaving a field of residual electricity in a 3x3 area for 9 seconds (also subject to change).
- Mobs hit directly by the bolt take 35 burn damage, mobs in the 3x3 area take 25. Both get disorientated for a bit and lose some stamina. Small robots (like beepsky) in the 3x3 area when it strikes will die instantly. Cyborgs just take regular damage. It also makes sparks which can set fuel tanks off and generally do what sparks do in this game.
- The residual electricity that's left over disorientates you if you move through it. They don't stack on tiles, casting a new lightning bolt on an overlapping tile will 'refresh' the tile and extend the life of it back up to 9 seconds. Therefore, it's slightly more efficient to spread them out if you want to maximize residual electricity in an area.
- Chaplains and those with SMES human are totally immune to both of these. Can walk straight through residue and take lightning bolts like it's nothing. Other wizards are also totally immune. The caster (you) aren't immune at all, so be careful! You cannot target tiles in the chapel or use it from inside the chapel at all.
- The spell you get recalls the wand, just like SoC. If it's used outside of your inventory it also leaves a 3x3 residual spark field from where it teleported *from*. In addition, it can be used from in your hand because...
- The recall spell also recharges the lightning bolts. The wand has a 3 charge maximum that's totally refilled when recalled. The recall is on a 20 second cooldown.
- Non-wizards can't pick it up, but aren't damaged by it. It does normal staff damage, not particularly impressive as a melee weapon at all. For melee, it's basically a fancy stick.
- It also functions as a regular staff for casting spells and such.
- Awesome lightning bolt sfx by TheMightyAltroll !
- Four states of charge animation, and a special red variant if admins var edit it to have more than three. Animations for firing and recharging.
- Functional wizard ring!

![unknown](https://user-images.githubusercontent.com/53125172/164874352-c13cad16-58f0-4484-ab5b-ab939835c5df.png)

![unknown](https://user-images.githubusercontent.com/53125172/164874361-3a97c8cb-a2f5-42a2-b14d-5b660779e18a.png)

(I recorded a longer video than this but it was too large for github so I'm using this one instead)

https://user-images.githubusercontent.com/53125172/164874443-c1c3897a-a396-44a8-b9e5-1aa7ade7b24d.mp4



## Why's this needed? <!-- Describe why you think this should be added to the game. -->

Wizard wands are cool. I think this has it's own niche and 'feel' to use. In my short testing it's fun to bait beepsky into range of it and just barely avoiding hitting yourself with lightning. It can be a neat area denial tool that doesn't *fully* deny an area, but covers a lot of ground and synergizes well with spells that let you move over the residual electricity you make. It's also ok-lethal, they either have to be still or get boxed in through clever marker placement by the wizard, and it takes a lot of charges to actually kill someone (it's used in tandem with other spells though). Summoning lightning bolts is very cool and I think people would enjoy doing it.

It seems like fun to cast three shots, then throw your wand at a spot and recall it strategically to maximize the residual electricity that makes. It's like an extra layer of tech that I hope makes it unique to use and makes people feel skilled.

Some notes on implementation. This is probably the most ambitious coding project I've done by myself, and I expect it to not really be perfect. Testing in-game has it working and I tried to account for as many weird buggy things I could (like limiting the range and stuff). The lightning bolt gets its own 64x128.dmi file because the only other file that would've fit it was 160x160 and when I was using that I noticed a substantial amount of lag when rendering the bolt for the first time. My understanding is that loading large dmis (as in, large pngs) is taxing and this is a Combat Antag Only Item, so I think it's important to not have it lag on first use. Either way, the lag was gone after moving it to it's own file. The code for the spell is essentially the same as the SoC but modified to allow for charging the wand. Unsure if that could've been done better.

also thanks to pali and arthur in discord for very useful pointers and tips! and to TheMightyAltroll for making the really cool lightning strike sound

## Changelog
<!-- If necessary, put your changelog entry below. Otherwise, please delete it.
Use however you want to be credited in the changelog in place of CodeDude.
Use (*) for major changes and (+) for minor changes. For example: -->

```changelog
(u)Flaborized & TheMightyAltroll
(*)Added a new wizard spell, 'Stave Of Thunder'. It can summon three lightning bolts to strike your foes with lingering disorientating electricity before needing to be recharged by recalling it. Awesome lightning bolt sound effect from TheMightyAltroll!
```

Redtrogdor commented:
I really like this, but would this render SoC obsolete? Not saying that this is too OP or anything, just proposing the idea that maybe SoC should be looked at and modified at some point. Maybe as more of a passive, defensive staff with minor buffs while this one is based more around active area denial and damage-dealing

flappybatpal commented:
> Wizards (including the user) aren't! 

Not sure on this with that big lingering AOE. Could be a real pain if another wizard hits you with it.

Flaborized commented:
> I really like this, but would this render SoC obsolete? Not saying that this is too OP or anything, just proposing the idea that maybe SoC should be looked at and modified at some point. Maybe as more of a passive, defensive staff with minor buffs while this one is based more around active area denial and damage-dealing

Maybe. I'm hoping that not having that strong melee weapon that stuns and kills in two ways thing going for it separates it enough. I think it might overtake SoC in the 'cool item' factor, which could lead to it being taken less. It'd be good if SoC was a bit cooler, but obviously out of scope of this PR.

> > Wizards (including the user) aren't!
> 
> Not sure on this with that big lingering AOE. Could be a real pain if another wizard hits you with it.

Agreed. I was initially worried that it might be a bit strong if used together with another wizard, but the more I think on it the more I realize that it's both good to encourage that kind of synergistic thinking and bad to make other wizards an active nuisance. Changed to make it not affect wizards that aren't the caster whatsoever (the bolt and the lingering effect)

Flaborized commented:
https://user-images.githubusercontent.com/53125172/164912612-8f2341d8-0eb4-4f06-9938-cb1f7f153598.mp4

Also added a sweet lightning sound affect from TheMightyAltroll. Very hefty sounding and cool

Flaborized commented:
Added sprites for charge levels, little energy orbs that show how many you have. Three = three charges, one = one, etc. Red variant for var edited wands that have more than three. Lowered the filesize of the sound to 29kb on request of virva (can go lower if need be) and added an existing spark noise to putting down targets. also, a wizard ring!

https://user-images.githubusercontent.com/53125172/164988328-7102d818-4eb2-4e26-8cf1-0fa4fd2d0e84.mp4

ZeWaka commented:
One thing that would be nice:
Currently the 'lighting zone' sprite looks very Tiled, if that makes sense. Maybe having some variant icon states would help break this up, or applying a mask around the edges to make it fade out, so it doesn't suddenly seem to just 'stop' at the end of the 3x3 tiles (probly. too much work).

ZeWaka commented on code/obj/item/misc_wizard_stuff.dm:
```diff
@@ -220,6 +220,85 @@
+				..()
+				return
+			else
+				boutput(user, "<span class='alert'>Your hand phases through the [name]'s handle, like it's not there...</span>")
```
idk - i think shocking for 1 damage or something would be nice, otherwise people might just keep clicking it and not seeing the message because there's no non-text feedback

ZeWaka replied:
ditto for below stuff

Flaborized replied:
Agreed. Added a new proc that's called in the place of all of these, which deals 1 second of disorientate and plays the classic sparks4.ogg . I wanted to keep it away from doing real damage to people trying to pick it up, I think this is good enough feedback to help clear up that it's not just lag or something. Also doesn't disorientate if you have SMES human of course

Flaborized commented on code/datums/abilities/wizard/summon_staff.dm:
```diff
@@ -90,3 +90,98 @@
+		..()
+
+		var/list/staves = list()
+		var/we_hold_it = 0
```
yea most of the summon staff stuff is copy/paste from the existing summon staff spell. I'll fix all these up later :eye:

Flaborized replied:
(and fixed)

Flaborized commented on code/obj/item/misc_wizard_stuff.dm:
```diff
@@ -220,6 +220,85 @@
+			thunder_charges -= 1
+			var/turf/T = get_turf(target)
+			var/obj/lightning_target/lightning = new/obj/lightning_target(T)
+			playsound(T, "sound/effects/electric_shock_short.ogg", 70, 1)
```
fixed (also fixed in a different spot that I did that by accident)

Flaborized commented on code/modules/admin/gimmick/wizard_rings.dm:
```diff
@@ -86,6 +86,28 @@ ABSTRACT_TYPE(/obj/item/clothing/gloves/ring/wizard)
+				created_staff.wizard_key = user?.mind.key
+
+		disposing()
+			created_staff = null
```
Yea that's currently how the SoC ring works. Keeps the staff around but removes the pointer to the ring. I wanted to keep the behavior consistent and leaving behind the staff isn't so bad.

Flaborized commented on code/datums/abilities/wizard/summon_staff.dm:
```diff
@@ -90,3 +90,98 @@
+			return 1
+
+		if (M.getStatusDuration("stunned") > 0 || M.getStatusDuration("weakened") || M.getStatusDuration("paralysis") > 0 || !isalive(M) || M.restrained())
+			boutput(M, __red("Not when you're incapacitated or restrained."))
```
Fixed alongside the others

Flaborized commented on code/datums/abilities/wizard/summon_staff.dm:
```diff
@@ -90,3 +90,98 @@
+				if (!(S in staves))
+					staves["[S.name] #[staves.len + 1] [ismob(S.loc) ? "carried by [S.loc.name]" : "at [get_area(S)]"]"] += S
+
+		switch (staves.len)
```
Committed this change (and adjusted others that were the same)

Flaborized commented on code/datums/abilities/wizard/summon_staff.dm:
```diff
@@ -90,3 +90,98 @@
+		var/list/staves = list()
+		var/we_hold_it = 0
+		for_by_tcl(S, /obj/item/staff/thunder)
+			if (M.mind && M.mind.key == S.wizard_key)
```
implemented this fix too

Flaborized commented on code/datums/abilities/wizard/summon_staff.dm:
```diff
@@ -90,3 +90,98 @@
+				if (we_hold_it != 0)
+					for (var/obj/item/staff/thunder/T in M.contents)
+						T.recharge_thunder()
+					boutput(M, __red("You charge your stave in your hand."))
```
The parent objects are called 'staff' but this one's named a stave (which is more or less the same thing), mostly to give it a name that's very easy to quickly identify if you're skimming chat (ie: all the words are unique from other similar items)

ZeWaka replied:
no like you say staff literally 3 lines below this
you also named it Staff of Thunder and use staff in the description

Flaborized commented on code/datums/abilities/wizard/summon_staff.dm:
```diff
@@ -90,3 +90,98 @@
+					return 0
+
+			if (1)
+				var/obj/item/staff/thunder/S2
```
agreed, changed to 'staff'

Flaborized commented on code/datums/abilities/wizard/summon_staff.dm:
```diff
@@ -90,3 +90,98 @@
+
+
+			if (2 to INFINITY)
+				var/t1 = input("Please select a staff to summon", "Target Selection", null, null) as null|anything in staves
```
sure, added

Flaborized commented on code/datums/abilities/wizard/summon_staff.dm:
```diff
@@ -90,3 +90,98 @@
+		if (!M)
+			return 1
+
+		if (M.getStatusDuration("stunned") > 0 || M.getStatusDuration("weakened") || M.getStatusDuration("paralysis") > 0 || !isalive(M) || M.restrained())
```
From what I can tell from looking at other abilities, they have individually defined 'incapacitation_check' procs. Like there's one for werewolf and one for vampire etc. Wizard doesn't already have one, but also only one currently existing spell (SoC) checks for any of this as far as I can tell. I could add a new proc, which would reduce copy/paste code and maybe futureproof new wiz spells a bit by giving them that to use if they want, but I'm unsure if there's something better to use instead. will sleep first :bee:

ZeWaka replied:
huh weird i thought i refactored these but i guess i didn't. huh.

Flaborized commented on code/datums/abilities/wizard/summon_staff.dm:
```diff
@@ -90,3 +90,98 @@
+					boutput(M, __red("You seem to have lost all magical abilities."))
+					return 0
+				if (M.wizard_castcheck(src) == 0)
+					return 0 // Has own user feedback.
```
Yeah, removed these entirely. Playtested the situations they'd interact with without them and seemingly nothing changed, so they're not needed it looks like

Flaborized commented:
> One thing that would be nice: Currently the 'lighting zone' sprite looks very Tiled, if that makes sense. Maybe having some variant icon states would help break this up, or applying a mask around the edges to make it fade out, so it doesn't suddenly seem to just 'stop' at the end of the 3x3 tiles (probly. too much work).

Agreed. I thought about and tried out having varying tile types early on but I thought it looked too noisy, since you're supposed to be able to really quickly identify them in combat and be able to easily tell what's going on. Very cluttery with more random looking patterns imo. I'd love to draw up and implement some like, edge-tiles for them to autotile with (since I don't like how they look so square) but I basically have zero idea how to code that whatsoever so I haven't :mortydead: More of a circle of electricity instead of an open set of tiles would be very nice

Flaborized commented:
Changed it from stave of thunder to staff of thunder, sound less weird and is better I think.

Sovexe commented on code/WorkInProgress/AnimationLibrary.dm:
```diff
@@ -1210,6 +1210,20 @@ proc/muzzle_flash_any(var/atom/movable/A, var/firing_angle, var/muzzle_anim, var
+	var/turf/target_turf = get_turf(target)
+	if (!target_turf)
+		return
+	var/obj/decal/lightning_bolt/lightning = new /obj/decal/lightning_bolt
+	lightning.set_loc(target_turf)
+	lightning.pixel_x = -4
+	SPAWN(0.7 SECONDS)
+		if (lightning)
+			qdel(lightning)
+	return
```
These tasks should probably be handled by the lightning bolt itself
```suggestion
	new /obj/decal/lightning_bolt(target_turf)
```

Sovexe commented on code/procs/lightning_bolt.dm:
```diff
@@ -0,0 +1,86 @@
+	plane = PLANE_NOSHADOW_ABOVE
+	icon = 'icons/effects/64x128.dmi'
+	icon_state = "lightning_1"
+
+	New()
+		icon_state = pick("lightning_1", "lightning_2")
+		..()
```
```suggestion
	pixel_x = -4

	New()
		..()
		flick(pick("lightning_1", "lightning_2"), src)
		SPAWN(0.7 SECONDS)
			qdel(src)
```

Sovexe commented on code/datums/abilities/wizard/summon_staff.dm:
```diff
@@ -90,3 +90,92 @@
+	requires_robes = 1
+
+	cast(mob/target)
+		if (!holder)
+			return 1
+
+		var/mob/living/M = holder.owner
+
+		if (!M)
+			return 1
+
+		if (M.getStatusDuration("stunned") > 0 || M.getStatusDuration("weakened") || M.getStatusDuration("paralysis") > 0 || !isalive(M) || M.restrained())
+			boutput(M, "<span class='alert'Not when you're incapacitated or restrained.")
+			return 1
```
no need to check for nulls step by step, we can just do it all at once. Also that ugly long if statement didn't stop pins
```suggestion
		var/mob/living/M = holder?.owner
		if (!ismob(M)) 
			return 1
		if (!can_act(M))
			boutput(M, "<span class='alert'Not when you're incapacitated or restrained.")
			return 1
```

Sovexe commented on code/datums/abilities/wizard/summon_staff.dm:
```diff
@@ -90,3 +90,92 @@
+					staves["[S.name] #[length(staves) + 1] [ismob(S.loc) ? "carried by [S.loc.name]" : "at [get_area(S)]"]"] += S
+
+		switch (length(staves))
+			if (-INFINITY to 0)
```
```suggestion
			if (0)
```

Sovexe commented on code/datums/abilities/wizard/summon_staff.dm:
```diff
@@ -90,3 +90,92 @@
+
+				var/obj/item/staff/thunder/staff = staves[t1]
+
+				if (!M || !ismob(M))
+					return 0
+				if (!staff || !istype(staff))
+					boutput(M, "<span class='alert'You were unable to summon your staff.")
+					return 0
+				if (M.getStatusDuration("stunned") > 0 || M.getStatusDuration("weakened") || M.getStatusDuration("paralysis") > 0 || !isalive(M) || M.restrained())
```
```suggestion
				if (!M)
					return 0
				if (!istype(staff))
					boutput(M, "<span class='alert'You were unable to summon your staff.")
					return 0
				if (!can_act(M))
```

Sovexe commented on code/obj/item/misc_wizard_stuff.dm:
```diff
@@ -220,6 +220,93 @@
+		START_TRACKING
+
+	disposing()
+		. = ..()
+		STOP_TRACKING
```
```suggestion
		STOP_TRACKING
		. = ..()
```

Sovexe commented on code/obj/item/misc_wizard_stuff.dm:
```diff
@@ -220,6 +220,93 @@
+		STOP_TRACKING
+
+	pixelaction(atom/target, params, mob/user, reach)
+		if(!IN_RANGE(user, target, 10))
```
```suggestion
		if(!IN_RANGE(user, target, WIDE_TILE_WIDTH / 2))
```

Sovexe reviewed (changes requested):
this is all for now, to tired to finish at the moment

virvatuli-gh reviewed (approved):
Oh yeah I forgot I was supposed to update you on the filesize request. 29kb is fine for an almost 3 second sound that relies on sounding properly oomphy. :thumbsup:

Sovexe reviewed (approved):
looks fine
